doc/dev/static_type_checking.md

```diff
@@ -0,0 +1,894 @@
+
+> Note that TypeAlias is backported to older versions of Python by using typing_extensions
+
+### Typing variadic arguments - *args and **kwargs
+
+Many of our client methods take a `**kwargs` and sometimes an `*args` argument.
+
+Consider the below example:
+
+```python
+def begin_operation(*args, **kwargs) -> None:
+    ...
+```
+
+Here, let's assume that `*args` accepts an arbitrary number of positional parameters which must be of type `str`. Let's
+also assume that `**kwargs` accepts operation-specific keyword arguments, including any `azure-core` specific keyword
+arguments. In this case, we can narrow the type for `*args` to `str`. For `**kwargs`, due to the complexity of the type,
+we will leave it as `Any`. Typed this looks like:
+
+```python
+from typing import Any
+
+
+def begin_operation(*args: str, **kwargs: Any) -> None:
+    ...
+```
+
+Note how mypy understands these types:
+
+```python
+from typing import Any
+
+
+def begin_operation(*args: str, **kwargs: Any) -> None:
+    reveal_locals()
+```
+
+```cmd
+main.py:4: note: Revealed local types are:
+main.py:4: note:     args: builtins.tuple[builtins.str]
+main.py:4: note:     kwargs: builtins.dict[builtins.str, Any]]"
+```
+
+`mypy` correctly infers `*args` as a `tuple[str]` and `**kwargs` as a `dict[str, Any]` so we do not need to type these
+as such in the code. If `*args` accepts more than just `str`, a `Union[]` type or `Any` can be used depending on how
+complex the type is.
+
+### Use AnyStr when your parameter accepts both str and bytes
+
+The `typing` module includes a pre-defined `TypeVar` named `AnyStr`. Its definition looks like this:
+
+```python
+from typing import TypeVar
+
+AnyStr = TypeVar('AnyStr', bytes, str)
+```
+
+`AnyStr` can be used in functions which accept or return both bytes and str.
+
+> Note: The `TypeVar` is a generic type which restricts `AnyStr` to `bytes` or `str`. More information on [TypeVar](https://docs.python.org/3/library/typing.html#typing.TypeVar)
+
+### Use typing.cast to help mypy understand a type
+
+Sometimes `mypy` will be unable to infer a type and may raise a false positive error. Using `typing.cast` is a way to
+tell `mypy` what the type of something is and is generally favored over using a `# type: ignore`.
+
+The below snippet shows an example of how `mypy` might need a little help to understand the parameter type specified in
+a union:
+
+```python
+from typing import Union, List, Optional, Any, cast
+
+
+def receive_deferred_messages(
+        self, sequence_numbers: Union[int, List[int]], *, timeout: Optional[float] = None, **kwargs: Any
+) -> List[ServiceBusReceivedMessage]:
+    if isinstance(sequence_numbers, int):
+        sequence_numbers = [sequence_numbers]
+    sequence_numbers = cast(List[int], sequence_numbers)  # mypy clarity
+```
+
+If you are using cast often, however, it might be an indication that the code should be otherwise written or refactored.
```

Comment:
Let's add a note that it's cost-free, it doesn't slow down the implementation at runtime